### DISCUSSION TITLE
Input inside Composite receives blur/focus on CompositeItem click

### DISCUSSION THREAD

**Discussion by vvo** [ariakit/ariakit#724] [Questions] [answered]
### Current behavior

- When using an input inside Composite, just like this isolated example does: https://codesandbox.io/s/composite-forked-xleis?file=/index.js
- trying to listen for input focus/blur to be able to show/hide CompositeItem list
- When you click on input: focus is received
- When you click outside of input: blur is received
- When input is focused
- When you click on an item: input receives blur then focus (two events)

It seems odd to receive both event, and it makes it impossible to show/hide the list of items. Because if you hide the list of item, then onClick on item is never fired (list hides first).

### Steps to reproduce the bug

Provide a repo or sandbox with the bug and describe the steps to reproduce it.

1. Open https://codesandbox.io/s/composite-forked-xleis?file=/index.js
2. Focus input
3. Click on an item
4. Input receives blur + focus

### Expected behavior

Since focus stays on input all long, it should not receive blur or focus on item click

### Possible solutions

-

### Environment

Please, run the command below inside your project directory.
```sh
$ npx envinfo --system --binaries --browsers --npmPackages "{react*,reakit*}"

npx: installed 1 in 2.845s

  System:
    OS: macOS 10.15.6
    CPU: (4) x64 Intel(R) Core(TM) i7-6567U CPU @ 3.30GHz
    Memory: 77.58 MB / 16.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.18.3 - ~/.nvm/versions/node/v12.18.3/bin/node
    Yarn: 1.22.4 - ~/.yarn/bin/yarn
    npm: 6.14.6 - ~/.nvm/versions/node/v12.18.3/bin/npm
  Browsers:
    Chrome: 84.0.4147.135
    Firefox: 73.0.1
    Safari: 13.1.2
  npmPackages:
    react: 16.13.1 => 16.13.1
    reakit: 1.2.2 => 1.2.2
```

[a comment by an automated account is not shown]

**Comment by diegohaz** [comment, marked as answer]
Hey @vvo! This behavior is better explained here (more at the end of the thread): https://twitter.com/diegohaz/status/1291005921712275456

When you use render props and you pass props like `onFocus` and `onBlur` directly to the underlying element (`input`), you're overriding the internal component `htmlProps` that might contain them as well. That's okay if you really intend to do it.

Since that doesn't seem to be your case, you have two options:

1. Check if `htmlProps.onFocus` and `htmlProps.onBlur` coming from the Composite component have called `event.stopPropagation()`:

```jsx
<Composite {...composite}>
  {htmlProps => (
    <input
      {...htmlProps}
      onFocus={event => {
        if (htmlProps.onFocus) {
          htmlProps.onFocus(event);
        }
        if (event.isPropagationStopped()) {
          // this event should be ignored
          return;
        }
        console.log("focus");
      }}
    />
  )}
</Composite>
```

2. Or, more simply, just pass `onFocus` to the `Composite` component instead of the underlying element. `Composite` will handle it for you:

```jsx
<Composite {...composite} onFocus={() => console.log("focus")}>
  {htmlProps => (<input {...htmlProps} />)}
</Composite>
```

That CodeSandbox may be misleading as there's a lot of props (`role`, `aria-haspopup` etc.) being passed to the underlying `input` element instead of the `Composite` component, whereas only `onKeyDown` and `onKeyUp` are really overwriting the default behavior. I just updated the original Composite sandbox to better reflect this.

Forked your CodeSandbox too: https://codesandbox.io/s/composite-forked-4jzuw?file=/index.js

**Comment by vvo** [comment]
Thanks for this! It worked to pass onFocus/onBlur directly on the Composite.
